### PR TITLE
[flang][cuda] Support memory cleanup at a return statement

### PR DESCRIPTION
**Pull request by khaki3**
We generate `cuf.free` and `func.return` twice if a return statement exists at the end of program.

```f90
program test
  integer, device :: a(10)
  return
end
```

```
% flang -x cuda test.cuf -mmlir --mlir-print-ir-after-all
error: loc("/path/to/test.cuf":3:3): 'func.return' op must be the last operation in the parent block
// -----// IR Dump After Fortran::lower::VerifierPass Failed () //----- //
```

Dumped IR:
```mlir
  "func.func"() <{function_type = () -> (), sym_name = "_QQmain"}> ({
...
    "cuf.free"(%5#1) <{data_attr = #cuf.cuda<device>}> : (!fir.ref<!fir.array<10xi32>>) -> ()
    "func.return"() : () -> ()
    "cuf.free"(%5#1) <{data_attr = #cuf.cuda<device>}> : (!fir.ref<!fir.array<10xi32>>) -> ()
    "func.return"() : () -> ()
}
...
```

The routine `genExitRoutine` in `Bridge.cpp` is guarded by `blockIsUnterminated()` to make sure that `func.return` is generated only at the end of a block. However, we redundantly run `bridge.fctCtx().finalizeAndKeep()` before `genExitRoutine` in this case, resulting in two pairs of `cuf.free` and `func.return`. This PR fixes `Bridge.cpp` by using `blockIsUnterminated()` to guard `finalizeAndKeep` as well.

### CODE REVIEW

File: flang/include/flang/Lower/StatementContext.h [hunk 1]
```diff
@@ -79,23 +79,26 @@ class StatementContext {
     }
   }
 
-  /// Make cleanup calls. Retain the stack top list for a repeat call.
+  /// Make a cleanup call. Retain the stack top list for a repeat call.
```
Comment (clementval) [inline flang/include/flang/Lower/StatementContext.h:82]: I think you can keep the original comment since we generate 0 to N cleanup calls.

**Review by clementval — COMMENTED** [review]
Thanks to work on this Matsu! This look ok to me. Just adding Jean and Val to have there feedback. There might be a case we forgot.

**Comment by clementval** [comment]
Can you add a test with multiple return like: 

```
program test
  integer, device :: a(10)
  logical :: l

  if (l) then
    return
  end if

  return
end
```

**Review by vdonaldson — APPROVED** [review]
This looks like an improvement to me as is.

I'm not familiar with OpenACC code. But for the record, if OpenACC regions are directives in the PFT, it might be possible to further simplify this code by subsuming the `openAccCtx` into the `activeConstructStack`, and/or possibly even integrating those with the `fctCtx`. That would be a little more ambitious though.

**Pull request merged**